src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,997 @@
+    //           1 0 1 0 0 0 1 1
+    new_lo |= bits._ones;
+    // In this case, new_lo may not always be a valid answer. This can happen
+    // if there is no bit up to first_violation that is 0 in both lo and zeros,
+    // i.e. tmp == 0. In such cases, alignment == 0 && lo == bits._ones. It is
+    // the only case when this function does not return a valid answer.
+    assert(lo < new_lo || new_lo == bits._ones, "overflow must return bits._ones");
+    return new_lo;
+  }
+}
+
+// Try to tighten the bound constraints from the known bit information. I.e, we
+// find the smallest value not smaller than lo, as well as the largest value
+// not larger than hi both of which satisfy bits
+// E.g: lo = 0010, hi = 1001
+// zeros = 0011
+// ones  = 0000
+// -> 4-aligned
+//
+//         0    1    2    3    4    5    6    7    8    9    10
+//         0000 0001 0010 0011 0100 0101 0110 0111 1000 1001 1010
+// bits:   ok   .    .    .    ok   .    .    .    ok   .    .
+// bounds:           lo                                 hi
+// adjust:           --------> lo                  hi <---
+template <class U>
+static AdjustResult<RangeInt<U>>
+adjust_bounds_from_bits(const RangeInt<U>& bounds, const KnownBits<U>& bits) {
+  U new_lo = adjust_lo(bounds._lo, bits);
+  if (new_lo < bounds._lo) {
+    // This means we wrapped around, which means no value not less than lo
+    // satisfies bits
+    return AdjustResult<RangeInt<U>>::make_empty();
+  }
+
+  // We need to find the largest value not larger than hi that satisfies bits
+  // One possible method is to do similar to adjust_lo, just with the other
+  // direction
+  // However, we can observe that if v satisfies {bits._zeros, bits._ones},
+  // then ~v would satisfy {bits._ones, bits._zeros}. Combine with the fact
+  // that bitwise-not is a strictly decreasing function, if new_hi is the
+  // largest value not larger than hi that satisfies {bits._zeros, bits._ones},
+  // then ~new_hi is the smallest value not smaller than ~hi that satisfies
+  // {bits._ones, bits._zeros}.
+  //
+  // Proof:
+  // Calling h the smallest value not smaller than ~hi that satisfies
+  // {bits._ones, bits._zeros}.
+  //
+  // 1. Since h satisfies {bits._ones, bits._zeros}, ~h satisfies
+  //   {bits._zeros, bits._ones}. Assume the contradiction ~h does not satisfy
+  //   {bits._zeros, bits._ones}, There can be 2 cases:
+  //   1.1. There is a bit in ~h that is 0 where the corresponding bit in ones
+  //     is 1. This implies the corresponding bit in h is 1. But this is
+  //     contradictory since h satisfies {bits._ones, bits._zeros}.
+  //   1.2. There is a bit in ~h that is 1 where the corresponding bit in zeros
+  //     is 1. Similarly, this leads to contradiction because h needs to
+  //     satisfy {bits._ones, bits._zeros}.
+  //
+  // 2. Assume there is a value k that is larger than ~h such that k is not
+  // larger than hi and k satisfies {bits._zeros, bits._ones}. As a result, ~k
```

Comment:
```suggestion
  // 2. Assume there is a value k that is larger than ~h such that k is not
  // larger than hi, i.e. ~h < k <= hi  and k satisfies {bits._zeros, bits._ones}. As a result, ~k
```
Just to help the reader a little.